### DISCUSSION TITLE
How do I generate PDF files programmatically?

### DISCUSSION THREAD

**Discussion by riccardostokker** [flxzt/rnote#238] [Development]
Hello!

First of all, I absolutely love the project. I'm using rnote daily to take my university notes, and it's a great example of GTK programming in Rust.

I'm currently writing a simple Rust program to sync files across my devices, and one of the tasks of this program would be converting *.rnote files to PDF files. I've been looking through the source code for a way to do this but I'm afraid that it's beyond my interpretation capabilities (I've just started learning Rust). What should I do to convert my rnote files to PDF programmatically?

Thank you,
Riccardo

**Comment by flxzt** [comment]
Hi, and thanks!

I would suggest looking at [engine.rs](https://github.com/flxzt/rnote/blob/main/rnote-engine/src/engine.rs) and [import.rs](https://github.com/flxzt/rnote/blob/main/rnote-engine/src/engine.rs) which contains methods `import_from_rnote_bytes_p1()` and `open_from_store_snapshot_p2()` for importing from a file.

To export, there is `export_doc_as_pdf_bytes()`.

Some of these methods give back a asynchronous channel receiver, so you would need to wrap them in [block_on](https://docs.rs/futures/latest/futures/executor/fn.block_on.html) for usage in a non-asynchronous program.

I hope this helps.

**Reply by riccardostokker** [reply]
Thank you very much!
